### PR TITLE
Panfrost drm WIP

### PR DESCRIPTION
**Pull request by bukinr**
All of these is needed for Panfrost driver operation with Mali T860 GPU.
Some changes are memory leak fixes.
I'm able to run sway and browsers (drm_kqfilter is needed for sway).
The Mali T860 has three hardware threads, so DRM scheduler is ported from Linux.
It works well with no memory or file destriptor leakages.
demo: https://twitter.com/onewilshire/status/1363792441871073280

### CODE REVIEW

**Comment by evadot** [comment]
Just a question before I start reading code, from which Linux version the scheduler have been taken ?
Currently this repo is synced with 5.6

**Comment by bukinr** [comment]
> Just a question before I start reading code, from which Linux version the scheduler have been taken ?
> Currently this repo is synced with 5.6
From linux master as of Nov-Dec 2020 I think, but I can sync with latest master or any branch

**Comment by evadot** [comment]
> > Just a question before I start reading code, from which Linux version the scheduler have been taken ?
> > Currently this repo is synced with 5.6
> 
> From linux master as of Nov-Dec 2020 I think, but I can sync with latest master or any branch

Ok, well if that works with what's in 5.6 that would ease future update to use that.

File: core/include/drm/drm_file.h [hunk 1]
```diff
@@ -352,6 +352,10 @@ struct drm_file {
 #if IS_ENABLED(CONFIG_DRM_LEGACY)
 	unsigned long lock_count; /* DRI1 legacy lock count */
 #endif
+
+	struct selinfo	drm_rsel;
```
Comment (evadot) [inline core/include/drm/drm_file.h:356]: Can you add some ifdef #FreeBSD here ?

File: core/include/drm/drm_file.h [hunk 2]
```diff
@@ -388,6 +392,7 @@ ssize_t drm_read(struct file *filp, char __user *buffer,
 		 size_t count, loff_t *offset);
 int drm_release(struct inode *inode, struct file *filp);
 __poll_t drm_poll(struct file *filp, struct poll_table_struct *wait);
+int drm_kqfilter(struct file *filp, struct knote *kn);
```
Comment (evadot) [inline core/include/drm/drm_file.h:395]: Same here

File: drmkpi/include/drmkpi/fs.h [hunk 3]
```diff
@@ -98,6 +98,7 @@ struct file_operations {
 	    struct file *file_out, loff_t pos_out, loff_t len,
 	    unsigned int remap_flags);
 	int (*fadvise)(struct file *, loff_t, loff_t, int);
+	int (*kqfilter)(struct file *, struct knote *);
```
Comment (evadot) [inline drmkpi/include/drmkpi/fs.h:101]: And here

**Review by evadot — CHANGES_REQUESTED** [review]

File: core/include/drm/drm_file.h (hunk 1 above)
Comment (bukinr) [reply core/include/drm/drm_file.h:357]: Added

File: core/include/drm/drm_file.h (hunk 2 above)
Comment (bukinr) [reply core/include/drm/drm_file.h:399]: Added

File: drmkpi/include/drmkpi/fs.h (hunk 3 above)
Comment (bukinr) [reply drmkpi/include/drmkpi/fs.h:102]: Added

**Comment by bukinr** [comment]
rework needed

**Pull request closed without merging**